backend/editor/entries.py

```diff
@@ -242,3 +242,38 @@ def update_node_children(entry, new_children_ids):
         result = session.run(query, {"id": entry, "child": child})
     
     return result
+
+def full_text_search(text):
+    """
+    Helper function used for searching a taxonomy
+    """
+    normalized_text = re.sub(r"([^A-Za-z0-9])", r"\\\1", text) # Escape special characters
+    text_query_exact = "*" + normalized_text + '*'
+    text_query_fuzzy = normalized_text + "~"
+
+    # Fuzzy search on two indexes
+    # Fuzzy search has more priority, since it matches more strings
+    query = f"""
+        CALL db.index.fulltext.queryNodes("nodeSearchIds", $textqueryfuzzy) YIELD node, score
+        WHERE score > 0.2
+        RETURN node.id
+        UNION
+        CALL db.index.fulltext.queryNodes("nodeSearchTags", $textqueryfuzzy) YIELD node, score
+        WHERE score > 0.2
+        RETURN node.id
+    """

```

Comment:
You can do all your queries in one go by introducing an "order by score" and multiplying the score by a constant (here I choose 5, but we make tweak more finely):


```suggestion
    query = f"""
        CALL db.index.fulltext.queryNodes("nodeSearchIds", $textqueryfuzzy)
        YIELD node, score as score_
        WHERE score_ > 0.2
        RETURN node.id, score_ * 5 as score
        ORDER BY score DESC
        UNION
        CALL db.index.fulltext.queryNodes("nodeSearchTags", $textqueryfuzzy)
        YIELD node, score as score_
        WHERE score_ > 0.2
        RETURN node.id, score_ * 5 as score
        ORDER BY score DESC
        CALL db.index.fulltext.queryNodes("nodeSearchIds", $textqueryexact) YIELD node, score
        WHERE score > 0.2
        RETURN node.id, score
        ORDER BY score DESC
        UNION
        CALL db.index.fulltext.queryNodes("nodeSearchTags", $textqueryexact) YIELD node, score
        WHERE score > 0.2
        RETURN node.id, score
        ORDER BY score DESC
    """

```

Reply:
a bit more complicated than I though !
Here is the real query after some test :-)

```suggestion
    query = """
        CALL {
                CALL db.index.fulltext.queryNodes("nodeSearchIds", $textqueryfuzzy)
                yield node, score as score_
                return node, score_ * 5 as score
            UNION
                CALL db.index.fulltext.queryNodes("nodeSearchTags",  $textqueryfuzzy)
                yield node, score as score_
                return node, score_ * 5 as score
            UNION
                CALL db.index.fulltext.queryNodes("nodeSearchIds",  $text_query_exact)
                yield node, score as score_
                return node, score_ as score
            UNION
                CALL db.index.fulltext.queryNodes("nodeSearchTags", $text_query_exact)
                yield node, score as score_
                return node, score_ as score 
        }
        with node.id as node, score

        RETURN node, sum(score) as score
        
        ORDER BY score DESC
    """

```

Reply:
And a snippet to test it in neo4j:
```
CALL {
            CALL db.index.fulltext.queryNodes("nodeSearchIds", "yogurt~")
            yield node, score as score_
            return node, score_ * 5 as score
            UNION
            CALL db.index.fulltext.queryNodes("nodeSearchTags",  "yogurt~")
            yield node, score as score_
            return node, score_ * 5 as score
            UNION
            CALL db.index.fulltext.queryNodes("nodeSearchIds",  "*yog*")
            yield node, score as score_
            return node, score_ as score
            UNION
            CALL db.index.fulltext.queryNodes("nodeSearchTags", "*yog*")
            yield node, score as score_
            return node, score_ as score 
        }
        with node.id as node, score

        RETURN node, sum(score) as score
        
        ORDER BY score DESC
```

Reply:
@alexgarel this doesn't work for the query string "meat". The score is skewed towards "en:fake-meat", but "en:meat" is the closest string.

Reply:
@alexgarel I think we need to give a little less priority to the `nodeSearchIds` index during fuzzy search, and use take the average score instead of the sum.
```
CALL {
            CALL db.index.fulltext.queryNodes("nodeSearchIds", "yogurt~")
            yield node, score as score_
            return node, score_ * 3 as score
            UNION
            CALL db.index.fulltext.queryNodes("nodeSearchTags",  "yogurt~")
            yield node, score as score_
            return node, score_ * 5  as score
            UNION
            CALL db.index.fulltext.queryNodes("nodeSearchIds",  "*yogurt*")
            yield node, score as score_
            return node, score_ as score
            UNION
            CALL db.index.fulltext.queryNodes("nodeSearchTags", "*yogurt*")
            yield node, score as score_
            return node, score_ as score 
        }
        with node.id as node, score

        RETURN node, sum(score)/count(score) as score
        
        ORDER BY score DESC
```
Both query strings "yogurt" and "meat" worked perfectly with this. Is this okay?

Reply:
Good to have tested more thoroughly.
your proposal this is perfectly fine to me.